src/libsync/discovery.cpp

```diff
@@ -0,0 +1,1218 @@
+            }
+            processFileAnalyzeLocalInfo(item, path, localEntry, serverEntry, dbEntry, _queryServer);
+        };
+
+        if (!localEntry.isValid()) {
+            // Check for renames (if there is a file with the same file id)
+            bool done = false;
+            bool async = false;
+            // This function will be executed for every candidate
+            auto renameCandidateProcessing = [&](const OCC::SyncJournalFileRecord &base) {
+                if (done)
+                    return;
+                if (!base.isValid())
+                    return;
+
+                if (base._type == ItemTypeVirtualFileDownload) {
+                    // Remote rename of a virtual file we have locally scheduled
+                    // for download. We just consider this NEW but mark it for download.
+                    item->_type = ItemTypeVirtualFileDownload;
+                    done = true;
+                    return;
+                }
+
+                // Some things prohibit rename detection entirely.
+                // Since we don't do the same checks again in reconcile, we can't
+                // just skip the candidate, but have to give up completely.
+                if (base._type != item->_type && base._type != ItemTypeVirtualFile) {
+                    qCInfo(lcDisco, "file types different, not a rename");
+                    done = true;
+                    return;
+                }
+                if (!serverEntry.isDirectory && base._etag != serverEntry.etag) {
+                    /* File with different etag, don't do a rename, but download the file again */
+                    qCInfo(lcDisco, "file etag different, not a rename");
+                    done = true;
+                    return;
+                }
+
+                // Now we know there is a sane rename candidate.
+                QString originalPath = QString::fromUtf8(base._path);
+
+                // Rename of a virtual file
+                if (base._type == ItemTypeVirtualFile && item->_type == ItemTypeFile) {
+                    // Ignore if the base is a virtual files
+                    return;
+                }
+
+                if (_discoveryData->_renamedItems.contains(originalPath)) {
+                    qCInfo(lcDisco, "folder already has a rename entry, skipping");
+                    return;
+                }
+
+                if (item->_type == ItemTypeFile) {
+                    csync_file_stat_t buf;
+                    if (csync_vio_local_stat((_discoveryData->_localDir + originalPath).toUtf8(), &buf)) {
+                        qCInfo(lcDisco) << "Local file does not exist anymore." << originalPath;
+                        return;
+                    }
+                    if (buf.modtime != base._modtime || buf.size != base._fileSize || buf.type != ItemTypeFile) {
+                        qCInfo(lcDisco) << "File has changed locally, not a rename." << originalPath;
```

Comment:
Some above set done=true; - these don't. It's inconsistent but probably doesn't cause bugs. I'm unconvinced what the best way of doing it would be.

Reply:
The logic here is unchanged from `_csync_detect_update`  (that's even your code :-)).
IMHO, indeed, none of the test above should have `done = true`  Should I change that now?